### ISSUE TITLE
AOC stops for long running  ECS task

### ISSUE THREAD

**Issue by hossain-rayhan** [aws-observability/aws-otel-collector#281]
**Whats wrong**
When I am running AOC using ECS with example [task-definition](https://github.com/aws-observability/aws-otel-collector/blob/main/examples/ecs/ecs-ec2-sidecar.json), it suddenly stops after running for a long time (6+ hours). 

**What I am expecting**
The task should run forever.

**What might be the issue**
We have a sample application which emits trace data. This container emits data for 6 hours or so and stops. As this container is marked as an essential container, it forces the task to stop. 

**What might be possible solutions**
I can think of two different solutions-
1. Mark the sample app container as non-essential. So, even if it stops, it won't kill our task.
2. Update the sample app container to stay alive even if it stops emitting trace data and keep it as an essential container.

**Comment by mjpowersjr** [comment]
I'm not sure if this is related, but I've been experimenting with running the collector as a ECS task as well. After running for a few hours the collector stopped responding to HTTP requests (client requests would hang open, and eventually time out).


As a stop-gap solution, I've introduced a health check to my collector based on the following example, and will be testing tonight.

https://medium.com/google-cloud/dockerfile-go-healthchecks-k8s-9a87d5c5b4cb

### Dockerfile
```dockerfile
FROM golang:latest AS build

COPY ./docker-health-check.go /opt
WORKDIR /opt

RUN CGO_ENABLED=0 \
    GOOS=linux \
    GOARCH=amd64 \
    go build -a -tags netgo docker-health-check.go

RUN ls -lah ./docker-health-check

FROM amazon/aws-otel-collector:v0.6.0

# AWS Image does not include curl for healthcheck!
COPY --from=build /opt/docker-health-check /

ADD ./otel-agent-config.yaml /etc/otel-agent-config.yaml

HEALTHCHECK \
    --interval=30s \
    --timeout=30s \
    --start-period=5s \
    --retries=3 \
    CMD [ "/docker-health-check", "http://localhost:13133" ]

# base image uses collector executable as imge Entrypoint
# this approach means all CLI args must be passed as
# the command to be run.
CMD [ "--config=/etc/otel-agent-config.yaml" ]
```

### docker-health-check.go
```go
// Derived closely from https://github.com/Soluto/golang-docker-healthcheck-example

package main

import (
	"fmt"
	"log"
	"net/http"
	"os"
)

func main() {
	if len(os.Args) < 2 {
		log.Fatal("Expected URL as command-line argument")
		os.Exit(1)
	}
	url := os.Args[1]
	fmt.Println(url)
	if _, err := http.Get(url); err != nil {
		os.Exit(1)
	}
}
```

**Comment by hossain-rayhan** [comment]
@mjpowersjr will wait for your result. Also, can you please share your OpenTelemetry config file.

**Comment by mjpowersjr** [comment]
Ok, I haven't had much luck - although I'm new to ECS, this might be a lack of understanding on my part.

1. My original health check didn't include a timeout in the Go code, although I believe the container runtime should have still timed out the healthcheck after 30s. I've edited the code in my above comment to reflect.
2. I've included my `otel-agent-config.yaml` as well, which is almost identical to the default.
3. I don't see a way in ECS to view the details of health checks. I'd like something equivalent to the following, but haven't found a way to do it:
```sh
docker inspect --format "{{json .State.Health }}" <container-id>
```
4. After including the health check, it seems that ECS is stuck constantly cycling my task every 10-20 minutes. Sometimes ECS also fails to pull my container image and restarting the task is delayed as a result...
Generally, my Service Events reads similar to the following:
```
2020-12-23 05:15:58 -0500   service otel-collector has started 1 tasks: task <task-id>.
2020-12-23 05:15:08 -0500   service otel-collector has stopped 1 running tasks: task <task-id>.
2020-12-23 05:15:08 -0500   service otel-collector task <task-id> failed container health checks.
2020-12-23 05:13:56 -0500   service otel-collector has started 1 tasks: task <task-id>.
2020-12-23 05:11:27 -0500   service otel-collector has started 1 tasks: task <task-id>.
2020-12-23 05:08:47 -0500   service otel-collector has started 1 tasks: task <task-id>.
2020-12-23 05:08:06 -0500   service otel-collector has stopped 1 running tasks: task <task-id>.
2020-12-23 05:08:06 -0500   service otel-collector task <task-id> failed container health checks.
```

One thing I don't understand is, the described health check seems to work fine in a local Docker environment. 

I think my next step will be to launch a container with the same image within a EC2 instances using Docker as a simple test. Obviously this is not a good final solution, but I think it will help me determine if I should be focusing on a ECS configuration issue or something with the collector.  

I still need to investigate how to provide IAM Role information to a container running within a EC2 instance, but I will need to circle back to this in a few days.

**Comment by mjpowersjr** [comment]
Maybe I should have noted in my last comment  - if someone who happens to know more about golang and/or ECS figures out what is going on before me - that sure would be a nice surprise. :-)

**Comment by hossain-rayhan** [comment]
Couple of points to be noted here-
1. I tried with our sample ECS task definition and default OpenTelemetry config file and marked the sample app (x-ray data emitter) as non-essential container. Everything worked fine and the task is running for three days. We have a known issue with our sample-app and working to fix it.
2. I am also not an expert with health check but would love to play with this. Would you mind sharing your full task definition?

**Comment by mjpowersjr** [comment]
@hossain-rayhan - I manually defined my cluster, service, and task (fargate based) via the ECS web interface. I'm happy to share more about that setup, I'd be more comfortable working with you out-of-band VS putting a lot of info in the public ticket. 

I also tried walking through the instructions and deploying the fargate sidecar example from this repo - but was not successful after a few attempts. Most commonly, it seems that the CF deployment hangs while attempting to create tasks. I don't see any specific errors in ECS or CF, I'm not sure how to troubleshoot that situation further.


![Screenshot from 2020-12-31 16-39-06](https://user-images.githubusercontent.com/133083/103426406-4ed23800-4b87-11eb-9f03-ee98140bd069.png)

**Comment by mjpowersjr** [comment]
After some trial and error, I was able to deploy a otel collector based on the `aws-otel-fargate-sidecare-deployment-cfn.yaml` template. I believe the previously mentioned timeout was being caused by either:
* the `aws-xray-data-emitter` task definition referencing a container image hosted in a different region
* or something to do with my VPC configuration `AssignPublicIp: DISABLED`

I was pretty tired at the time (and although I know better) I started changing more than one part of the template simultaneously until I had a successful deployment. FWIW, Cloud Formation never presented me with a meaningful reason for the deployment timeouts. 

----

The newly deployed collector appears to have run for longer, but ultimately ended up in a similar state as I originally described. Traces stopped flowing to X-Ray, and a simple curl request to the collector hangs open. This time I discovered log files in CloudWatch:

[otel-collector-error-log.txt](https://github.com/aws-observability/aws-otel-collector/files/5767970/otel-collector-error-log.txt)

**Comment by hossain-rayhan** [comment]
From my description, we followed the 2nd option- updated the sample app to keep the container alive even if it does not emit data. @JohnWu20 updated the sample-app last week. He tested it and the collector is running for 3 days which resolves my concern for opening this issue. 

Now, we still have one more issue to look-up for @mjpowersjr's case. If @mjpowersjr tried (yesterday) with the latest sample-app, then we need to investigate more to fix it. 
Though, I am not super clear, from the above attached log it feels like something wrong with the `aws-xray` exporter. I am not sure if we want to close this one and open a new issue but we need to try reproducing it first.

**Comment by JohnWu20** [comment]
Thanks @mjpowersjr for your feedback. 

For the cloudformation, if you couldn't see any errors which means that there is no error for the template. Then you can go to your cluster. You can see there is a service running in your services part like this:
![Screen Shot 2021-01-05 at 11 55 59 AM](https://user-images.githubusercontent.com/59711343/103692675-fc10cb80-4f4c-11eb-9f74-d99686bdcf42.png)
It's a service which will start your ECS task. If there is something wrong with your task, it will start a new one automatically. So if there is anything wrong with your setting or images, you can go to the tasks section to check the stopped tasks. You may find the reason for in the stopped task.

aws-xray-data-emitter task uses an ECR image, it's public and no regional restrictions.

And for the long run error you met, would you please upload the log files for the Xray emitter?  And I am trying to reproduce the error.

**Comment by mjpowersjr** [comment]
@JohnWu20 - I've checked the *Logs* tab for my `aws-otel-sidecar-service` and corresponding task, in the ECS web interface - both simply display *No results*. As far as I can tell, the only relevant feedback seems to be routed to CloudWatch. 

I spent some time tonight building out a simple proof of concept that demonstrates the issue using docker + docker-compose locally. Based off your feedback and research tonight - I believe that the number of open network sockets is growing indefinitely. It's not clear to me where the problem lies, but I'm starting to suspect it might be a issue with either `opentelemetry-js` or the `opentelemetry-collector` HTTP receiver.  (or some combination?). 

If the issue described needs to move to a new ticket or different project I'm happy to follow along wherever appropriate. Thanks again for everyone's help!

### Demo:
https://github.com/mjpowersjr/otel-collector-xray-test

**Comment by JohnWu20** [comment]
Thanks @mjpowersjr!
Based on your description, I found that you used the otlp receiver and xray exporter. OTLP receiver will listen to the 55681 port for http request, and Xray receiver will listen to 2000 UDP port. 
For our [ECS](https://aws-otel.github.io/docs/setup/ecs) instruction, our sample app xray trace data emitter will send xray segments to UDP port 2000 which Xray receiver will listen to. I tested it again yesterday, and it is still running now. So I think this issue opened in this repo is resolved.
And I believe the issue you described should come from either otlp receiver or Xray exporter. 
If you agree with me, I would suggest you open an issue in this [repo.](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues).

**Comment by mjpowersjr** [comment]
tl;dr - Yes I agree this ticket can be closed from my perspective.
----

After spending some more time testing different configurations, I think I've narrowed down the problem to either the otel collector http receiver or the opentelemetry-js http exporter. More specifically, I think the exporter is keeping connections open and not releasing or reusing them, causing the number of connections to grow over time. It also seems like the http receiver could better manage connections & timeouts to prevent accidental DoS scenarios.

As far as I can tell, a direct otel Xray exporter doesn't exist for javascript currently. I did run some tests using the otlp grpc exporter in my test repo, and the number of connections remained stable over time.

I did reach out to some of the otel community on Gitter, and i'm happy to file a follow on ticket(s) in some of the other repositories this week. Thanks to everyone for your time & help narrowing down the problem!
